### PR TITLE
Domain Management i1: Allow updating wpcom email address from bulk update contact form

### PR DESCRIPTION
**Pull request by vykes-mac**
This PR refactors the update wpcom email logic so that it can be executed on both bulk and non-bulk contact information updates if the user chooses the checkbox for that option.
Closes https://github.com/Automattic/dotcom-forge/issues/3652

## Proposed Changes

* Extract email update logic into reusable function.

## Testing Instructions

* `/domains/manage?flags=domains/bulk-actions-contact-info-editing`
* Select a domain then `edit contact information` button
* change email address and check the Apply` contact update to My Account email.` checkbox
* Verify you receive an email with information about updating your wpcom email address.
## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/domains/domain-management/edit-contact-info-page/bulk-edit-contact-info-page.tsx
Comment (zaguiini) [inline client/my-sites/domains/domain-management/edit-contact-info-page/bulk-edit-contact-info-page.tsx:143]: Can we remove this leftover?

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (zaguiini) [inline client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx]: I'd love if we could move this logic outside of the component and just call the thunk from this method (e.g., `this.props.updateWpcomEmail` and then everything is handled through Redux.)

File: client/my-sites/domains/domain-management/edit-contact-info-page/bulk-edit-contact-info-page.tsx
Comment (vykes-mac) [reply client/my-sites/domains/domain-management/edit-contact-info-page/bulk-edit-contact-info-page.tsx:143]: it's removed

Comment (zaguiini) [reply client/my-sites/domains/domain-management/edit-contact-info-page/bulk-edit-contact-info-page.tsx:143]: Oh, xD

Sorry!

**Review by zaguiini — APPROVED** [review]
Works well for me. Let's wait for another review before merging this one. cc @flootr do you have any thoughts about this? I know you worked on the settings section in the past.

File: client/state/user-settings/thunks/create-user-email-pending-update.js
Comment (p-jackson) [inline client/state/user-settings/thunks/create-user-email-pending-update.js]: I notice that we weren't previously displaying the error returned from the server. I initially got an error message saying "A fresh access token must be used to query information about the current user." I'm not sure whether that extra message returned by the API that useful for a user. How would I refresh an access token?

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (p-jackson) [inline client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx:394]: I think it's a little confusing that even though the `onSubmitButtonClick` prop has returned `'cancel'` the component isn't immediately canceling it's own processing.

The doc comment for the `onSubmitButtonClick` prop says returning `'cancel'` will cancel the default handling.

Given that `onSubmitButtonClick` is passed the value of `updateWpcomEmail` I think the bulk update page's submit button handler should be able to decide whether or not to update the wpcom email and dispatch the `createUserEmailPendingUpdate` action itself. By returning `'cancel'` it's saying we've handled everything that needs to be handled.

It doesn't change anything about how this PR runs, but I think it'll be clearer when someone revisits this code down the track.

**Comment by flootr** [comment]
> @flootr do you have any thoughts about this? I know you worked on the settings section in the past.

I only worked on these in the sense of frameworks and I don't have any objections about the code I'm seeing in that regard. Thanks for looping me in 👍🏻

File: client/state/user-settings/thunks/create-user-email-pending-update.js
Comment (zaguiini) [reply client/state/user-settings/thunks/create-user-email-pending-update.js]: In this case it's not helpful, that's true. But there are cases where the e-mail is already been used, and that case is returned by the API. Maybe if there's a way to get it through a constant, we could display this specific error.

File: client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
Comment (zaguiini) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx:394]: Do you mean you'd like to lift this decision to the component that is using `EditContactInfoFormCard`? The problem is there are two instances, so we'd need to call this in two different places. Plus, there's some logic that prevents or allows the calling of the update settings endpoint too.

I think one thing that can be done is to move the `this.updateWpcomEmail();` call outside the `onSubmitButtonClick` conditional:

```diff
diff --git a/client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx b/client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
index 540dbd73ad..5dbe0d2384 100644
--- a/client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
+++ b/client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx
@@ -385,13 +385,15 @@ class EditContactInfoFormCard extends Component {
 	};
 
 	handleSubmitButtonClick = ( newContactDetails ) => {
+		this.updateWpcomEmail( newContactDetails, this.state.updateWpcomEmail );
+
 		if ( this.props.onSubmitButtonClick ) {
 			const result = this.props.onSubmitButtonClick(
 				newContactDetails,
 				this.state.transferLock,
 				this.state.updateWpcomEmail
 			);
-			this.updateWpcomEmail( newContactDetails, this.state.updateWpcomEmail );
+
 			if ( result === 'cancel' ) {
 				return;
 			}

```

Like this, it will be clearer that the email update is an independent action to the submit button click handler. And well, it is.

Comment (p-jackson) [reply client/my-sites/domains/domain-management/edit-contact-info/form-card.jsx:394]: I see. Maybe my original idea of allowing components who use `EditContactInfoFormCard` to cancel processing just isn't a good fit anymore.

My intention behind doing it this way was that I didn't want to make a lot of changes to `EditContactInfoFormCard`. I thought I could mostly leave it as it was and then do any of the things needed for bulk updates in the parent component. So "we'd need to call this in two different places" as you say was an intentional trade off.

But as we've had to add more customisations to `EditContactInfoFormCard` (it even has a `bulkEdit` prop now) that intention doesn't make much sense any more. We could have all the logic in `handleSubmitButtonClick` now, calling the bulk update endpoints. And then it'd obviously just call `updateWpcomEmail` too.

Disabling the submit button while bulk updating also doesn't work correct at the moment. It'd fix that problem too.

So I take it back, I don't think we need to move the `updateWpcomEmail` out of this component any more.
We also don't have to make a refactor in this PR either. But it's good to figure it out in my head :)

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8969021

Hi @vykes-mac, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `There was a problem updating your WordPress.com account email: %(error)s`

Thank you in advance!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8969021).

**Pull request merged**
